Instant search

hughmiranda94 opened the pull request.

juristr reviewed (changes requested):
This PR needs some improvement.

Try to use more of a reactive approach. Using variables and `setTimeouts` can lead to complicated and buggy code.

Also make sure to address this requirement specified in the task description:

> Be sure not to spam the API with too many calls -- no more than one request every 500 ms.

hughmiranda94 commented:
@juristr I've updated this PR as well to have a more reactive approach and I have implemented the e2e test correctly. Let me know your feedback regarding these changes.

juristr commented on libs/books/feature/src/lib/book-search/book-search.component.ts:
```diff
@@ -34,9 +39,16 @@ export class BookSearchComponent implements OnInit {
+    const typing$ = fromEvent(this.searchInput.nativeElement, 'input');
+    typing$.pipe(debounceTime(500)).subscribe(value => this.searchBooks());
+    typing$.subscribe(value => this.isSearching = true);
+  }
```
👍 for `debounceTime(...)`, however you shouldn't subscribe multiple times to the same observable.

Also, instead of `fromEvent(...)` you can make use of Angular's ReactiveForm reactivity. It already has features to subscribe to form update which will give you an observable which you can then chain as you did in your example.

Could you adjust the example to do that?

hughmiranda94 replied:
I have updated the PR, could you please check it out?

juristr replied:
cool 👍 . Looks better, I'll give it a try later and let you know

juristr commented:
The changes look good to me now. However if I run the project build I'm getting

![image](https://user-images.githubusercontent.com/542458/94618967-57775180-02ac-11eb-845d-c1859303ca1b.png)

As a result of this build failure also the tests + e2e fail

hughmiranda94 commented:
@juristr I totally missed removing this class extension when I removed ngAfterViewInit on the past commit, thanks again for catching that! I've updated the PR, it should be good now

juristr commented:
LGTM now 👍
